seqr/views/utils/permissions_utils.py

```diff
@@ -150,21 +152,37 @@ def _get_analyst_projects():
     return ProjectCategory.objects.get(name=ANALYST_PROJECT_CATEGORY).projects.all()
 
 
-def get_projects_user_can_view(user):
-    if user_is_data_manager(user):
-        return Project.objects.all()
+def get_project_guids_user_can_view(user):
+    cache_key = 'projects__{}'.format(user)
+    project_guids = safe_redis_get_json(cache_key)
+    if project_guids is not None:
+        return project_guids
 
+    is_data_manager = user_is_data_manager(user)
+    if is_data_manager:
+        projects = Project.objects.all()
+    else:
+        projects = get_local_access_projects(user)
+
+    project_guids = [p.guid for p in projects.only('guid')]
+    if is_anvil_authenticated(user) and not is_data_manager:
+        workspaces = ['/'.join([ws['workspace']['namespace'], ws['workspace']['name']]) for ws in
+                      list_anvil_workspaces(user)]
+        project_guids += [p.guid for p in Project.objects.filter(workspace_name__isnull=False).exclude(
+            workspace_name='').exclude(guid__in=project_guids).annotate(
+            workspace=Concat('workspace_namespace', Value('/'), 'workspace_name')).filter(
+            workspace__in=workspaces).only('guid')]
```

Comment:
Doing filtering first then annotating will help for now when most of the projects haven't been associated with anvil workspaces. But, in the future, when more projects have workspaces configured, the annotation could cost more performance loss. I still prefer to add a field for combining the namespace and the name of the workspace. But, we could leave it as is and change it later based on further performance evaluation.

Reply:
Changing this now seems like a seriously premature optimization. This type of annotation is still pretty quick until you get up into the thousands of projects probably. The issue we were running into with the annotations today is because those annotations include joins with other tables, which can get slow. I will leave this as is, and we can revisit this idea in the future if we start to notice performance issues

Reply:
Okay